### DISCUSSION TITLE
Issue building armv5te-linux-gnueapi with different glibc version

### DISCUSSION THREAD

**Discussion by RedIODev** [cross-rs/cross#1152] [Q&A] [answered]
I'm trying to build a cdylib for armv5te-linux-gnueapi with glibc 2.24. But the installed glibc version on the default cross docker image for that target is 2.28 making the resulting library incompatible with the target Maschine (update glibc is not possible). I have tried various methods like adding a symlink asm to set the maximum glibc version or created a pre-execution script that installs glibc 2.24 and sets the environment to use it. Without any luck. It is a hard requirement to be a dynamic library and musl does not support it. I'm not very familiar with either cross or docker. If you have any idea on how to get this working please help.

**Comment by Alexhuszagh** [comment, marked as answer]
We've never supported anything under Ubuntu 18.04 for this target, but if you use Ubuntu 16.04, it comes with glibc 2.23 for the armel targets. Your Dockerfile then would be something like this:

```Dockerfile
FROM ubuntu:16.04
ARG DEBIAN_FRONTEND=noninteractive

COPY common.sh lib.sh /
RUN /common.sh

COPY cmake.sh /
RUN /cmake.sh

COPY xargo.sh /
RUN /xargo.sh

RUN apt-get update && apt-get install --assume-yes --no-install-recommends \
    g++-arm-linux-gnueabi \
    crossbuild-essential-armel \
    libc6-dev-armel-cross

COPY deny-debian-packages.sh /
RUN TARGET_ARCH=armel /deny-debian-packages.sh \
    binutils \
    binutils-arm-linux-gnueabi

# Qemu is disabled since we've changed the scripts to require newer Python versions.
#COPY qemu.sh /
#RUN /qemu.sh arm

COPY qemu-runner base-runner.sh /

ENV CROSS_TOOLCHAIN_PREFIX=arm-linux-gnueabi-
ENV CROSS_SYSROOT=/usr/arm-linux-gnueabi
ENV CARGO_TARGET_ARMV5TE_UNKNOWN_LINUX_GNUEABI_LINKER="$CROSS_TOOLCHAIN_PREFIX"gcc \
    CARGO_TARGET_ARMV5TE_UNKNOWN_LINUX_GNUEABI_RUNNER="/qemu-runner arm" \
    AR_armv5te_unknown_linux_gnueabi="$CROSS_TOOLCHAIN_PREFIX"ar \
    CC_armv5te_unknown_linux_gnueabi="$CROSS_TOOLCHAIN_PREFIX"gcc \
    CXX_armv5te_unknown_linux_gnueabi="$CROSS_TOOLCHAIN_PREFIX"g++ \
    BINDGEN_EXTRA_CLANG_ARGS_armv5te_unknown_linux_gnueabi="--sysroot=$CROSS_SYSROOT" \
    QEMU_LD_PREFIX="$CROSS_SYSROOT" \
    RUST_TEST_THREADS=1 \
    PKG_CONFIG_PATH="/usr/lib/arm-linux-gnueabi/pkgconfig/:${PKG_CONFIG_PATH}"
```

Save that file as `Dockerfile.armv5te-unknown-linux-gnueabi-cross` in the `docker` subdirectory of the cloned `cross` repository, then run `cargo build-docker-image armv5te-unknown-linux-gnueabi-cross` and then you can use this image by exporting the environment variable to `CROSS_TARGET_ARMV5TE_UNKNOWN_LINUX_GNUEABI_IMAGE=ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:local` before running cross. So, for example:

```bash
export CROSS_TARGET_ARMV5TE_UNKNOWN_LINUX_GNUEABI_IMAGE=ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:local
cross build --target armv5te-unknown-linux-gnueabi
```

Note that you will need most of cross's build scripts, so it is highly recommended you use the above workflow with `build-docker-image` rather than attempting a custom image with our pre-build hooks.

**Reply by RedIODev** [reply]
Docker cold not build the image:
```
[+] Building 31.6s (8/16)
 => [internal] load build definition from Dockerfile.armv5te-unknown-linux-gnueabi-cross                               0.2s 
 => => transferring dockerfile: 1.40kB                                                                                 0.1s 
 => [internal] load .dockerignore                                                                                      0.2s 
 => => transferring context: 234B                                                                                      0.1s 
 => [internal] load metadata for docker.io/library/ubuntu:16.04                                                       27.4s 
 => ERROR importing cache manifest from ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:main                      2.9s 
 => CACHED [ 1/11] FROM docker.io/library/ubuntu:16.04@sha256:1f1a2d56de1d604801a9671f301190704c25d604a416f59e03c04f5  0.0s 
 => [internal] load build context                                                                                      0.3s 
 => => transferring context: 10.65kB                                                                                   0.2s
 => [ 2/11] COPY common.sh lib.sh /                                                                                    0.0s
 => ERROR [ 3/11] RUN /common.sh                                                                                       0.6s 
------
 > importing cache manifest from ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:main:
------
------
 > [ 3/11] RUN /common.sh:
#0 0.569 /usr/bin/env: 'bash\r': No such file or directory
------
ERROR: failed to solve: executor failed running [/bin/sh -c /common.sh]: exit code: 127
Error: 
   0: `docker buildx build --platform linux/amd64 --build-arg 'CROSS_TARGET_TRIPLE=ARMV5TE_UNKNOWN_LINUX_GNUEABI_CROSS' --load --pull --cache-from 'type=registry,ref=ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:main' --tag ghcr.io/cross-rs/armv5te-unknown-linux-gnueabi-cross:local --label 'org.cross-rs.for-cross-target=armv5te-unknown-linux-gnueabi-cross' --label 'org.cross-rs.runs-with=x86_64-unknown-linux-gnu' --file /mnt/c/Users/RedFace/.vscode/source/cross/docker/Dockerfile.armv5te-unknown-linux-gnueabi-cross --progress auto .` failed with exit status: 1

Location:
   /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/convert/mod.rs:552

Warning: call to docker failed
Suggestion: is `buildx` available for the container engine?
Note: disable the `buildkit` dependency optionally with `CROSS_CONTAINER_ENGINE_NO_BUILDKIT=1`
```

**Reply by Emilgardis** [reply]
~~Have you tried the suggestion in the error message?~~

edit: nvm, this seems to be another issue, buildx is available, the dockerfile is just failing due to newlines

**Reply by Emilgardis** [reply]
it seems you have the cross repo checked out with windows newlines, see https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings?platform=linux for how to solve it


sidenote: we should fix this in `.gitattributes`

**Reply by RedIODev** [reply]
I will wrap this up in 1 answer in my stackoverflow question about this topic. [https://stackoverflow.com/questions/74492239/cross-compile-shared-library-for-armv5te-unknown-linux-gnueabi-rust](https://stackoverflow.com/questions/74492239/cross-compile-shared-library-for-armv5te-unknown-linux-gnueabi-rust) 
Of cause I will credit both of you and link this discussion.

**Reply by Alexhuszagh** [reply]
Glad to hear it works. I've updated your comment just to fix the URL btw, no other changes were made.
